File: 0070-climbing-stairs/0070-climbing-stairs.java
// TODO: 점화식을 먼저 세우자

/*
    dp[1] = 1 (step1)
    dp[2] = 2 (step1, step2)
    dp[3] = 3 = d[i-1] + dp[i-2]
    
    dp[4] = 5
*/        

class Solution {
    public int climbStairs(int n) {
        if (n <= 2) return n;

        int[] dp = new int[n+1];
        dp[1] = 1;
        dp[2] = 2;

        for (int i = 3; i <= n; i++) {
            dp[i] = dp[i-1] + dp[i-2];
        }

        return dp[n];
    }
}

File: 0070-climbing-stairs/0070-climbing-stairs.kt
class Solution {
    private val cache = mutableMapOf<Int, Int>()
    
    fun climbStairs(n: Int, count: Int = 0): Int {
        if (n == 0 || n == 1) return count + 1
        if (cache.containsKey(n)) return cache.get(n)!!
        
        val answer1 = climbStairs(n - 1, count)
        cache[n-1] = answer1
        
        val answer2 = climbStairs(n - 2, count)
        cache[n-2] = answer2
        
        return answer1 + answer2
    }
}